client/network/sync/src/warp.rs

```diff
@@ -104,51 +116,111 @@ where
 					Ok(VerificationResult::Complete(new_set_id, _, header)) => {
 						log::debug!(target: "sync", "Verified complete proof, set_id={:?}", new_set_id);
 						self.total_proof_bytes += response.0.len() as u64;
-						let state_sync = StateSync::new(self.client.clone(), header, false);
-						self.phase = Phase::State(state_sync);
+						self.phase = Phase::TargetBlock(header);
 						WarpProofImportResult::Success
 					},
 				}
 			},
 		}
 	}
 
+	/// Import the target block body.
+	pub fn import_target_block(&mut self, block: BlockData<B>) -> TargetBlockImportResult {
+		match &mut self.phase {
+			Phase::WarpProof { .. } | Phase::State(_) => {
+				log::debug!(target: "sync", "Unexpected target block response");
+				TargetBlockImportResult::BadResponse
+			},
+			Phase::TargetBlock(header) =>
+				if let Some(block_header) = &block.header {
+					if block_header == header {
+						if block.body.is_some() {
+							let state_sync = StateSync::new(
+								self.client.clone(),
+								header.clone(),
+								block.body,
+								false,
+							);
+							self.phase = Phase::State(state_sync);
+							TargetBlockImportResult::Success
+						} else {
+							log::debug!(target: "sync", "Importing target block failed: no body.");
+							TargetBlockImportResult::BadResponse
+						}
+					} else {
+						log::debug!(
+							target: "sync",
+							"Importing target block failed: different header.",
+						);
+						TargetBlockImportResult::BadResponse
+					}
+				} else {
+					log::debug!(target: "sync", "Importing target block failed: no header.");
+					TargetBlockImportResult::BadResponse
+				},
+		}
+	}
+
 	/// Produce next state request.
 	pub fn next_state_request(&self) -> Option<StateRequest> {
 		match &self.phase {
 			Phase::WarpProof { .. } => None,
+			Phase::TargetBlock(_) => None,
 			Phase::State(sync) => Some(sync.next_request()),
 		}
 	}
 
 	/// Produce next warp proof request.
-	pub fn next_warp_poof_request(&self) -> Option<WarpProofRequest<B>> {
+	pub fn next_warp_proof_request(&self) -> Option<WarpProofRequest<B>> {
 		match &self.phase {
-			Phase::State(_) => None,
 			Phase::WarpProof { last_hash, .. } => Some(WarpProofRequest { begin: *last_hash }),
+			Phase::TargetBlock(_) => None,
+			Phase::State(_) => None,
+		}
+	}
+
+	/// Produce next target block request.
+	pub fn next_target_block_request(&self) -> Option<(NumberFor<B>, BlockRequest<B>)> {
+		match &self.phase {
+			Phase::WarpProof { .. } => None,
+			Phase::TargetBlock(header) => {
+				let request = BlockRequest::<B> {
+					id: 0,
+					fields: BlockAttributes::HEADER | BlockAttributes::BODY,
```

Comment:
This should probably aso request and import `JUSTIFICATION`. Even though warp proof might have it in some form, it is not passed along with the header when validatoin is complete. I think it would not hurt to make sure it is imported here.

Reply:
Done.